lib/spack/spack/build_environment.py

```diff
@@ -59,6 +59,11 @@
 SPACK_DEBUG_LOG_DIR    = 'SPACK_DEBUG_LOG_DIR'
 
 
+# Platform-specific library suffix.
+dso_suffix = 'dylib' if sys.platform == 'darwin' else 'so'
```

Comment:
This is definitely not intended to solve the whole problem.  It's a first step and intended to get duplicate logic out of the package files.